File: Sigmath/Sigmath/Sigmath.ConvertedToRuby/calculus/calculus.rb
require "mscorlib"
require "System, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"
require "System.Collections.Generic, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"

module Sigmath
	class Calculus
		def Calculus.Derivative(f, x)
			dx = 1e-9 # small increment to find the derivative
			y1 = Calculus.f(x)
			y2 = Calculus.f(x + dx)
			derivative = (y2 - y1) / dx
			return derivative
		end

		def Calculus.Integral(f, a, b, n)
			h = (b - a) / n
			integral = 0
			i = 1
			while i < n
				x = a + i * h
				integral += Calculus.f(x)
				i += 1
			end
			integral += (Calculus.f(a) + Calculus.f(b)) / 2
			integral *= h
			return integral
		end

		def Calculus.Limit(f, g, x)
			maxIterations = 1000
			tolerance = 1e-9
			fx = Calculus.f(x)
			gx = Calculus.g(x)
			limit = fx / gx
			i = 0
			while i < maxIterations
				fx1 = Calculus.Derivative(f, x)
				gx1 = Calculus.Derivative(g, x)
				newLimit = fx1 / gx1
				if Basic.Abs(newLimit - limit) < tolerance then
					return newLimit
				end
				limit = newLimit
				x += 1e-9
				i += 1
			end
			raise Exception.new("Limit does not exist or cannot be computed with L'Hopital's Rule.")
		end

		def Calculus.FindRoot(f, a, b, epsilon)
			if Calculus.f(a) * Calculus.f(b) > 0 then
				raise ArgumentException.new("Function must have opposite signs at interval endpoints")
			end
			c = (a + b) / 2
			while Math.Abs(Calculus.f(c)) > epsilon
				if Calculus.f(a) * Calculus.f(c) < 0 then
					b = c
				else
					a = c
				end
				c = (a + b) / 2
			end
			return c
		end

		def Calculus.Factorial(n)
			result = 1
			i = 2
			while i <= n
				result *= i
				i += 1
			end
			return result
		end

		def Calculus.PartialDerivative(f, index, h)
			return 			xPlus = x.Clone()
			xPlus[index] += h
			xMinus = x.Clone()
			xMinus[index] -= h
			return (Calculus.f(xPlus) - Calculus.f(xMinus)) / (2 * h)

		end

		def Calculus.Gradient(f, h)
			return 			gradient = Array.CreateInstance(System::Double, x.Length)
			i = 0
			while i < x.Length
				gradient[i] = (x)
				i += 1
			end
			return gradient

		end

		def Calculus.Hessian(f, h)
			return 			n = x.Length
			hessian = Array.CreateInstance(System::Double, n, n)
			i = 0
			while i < n
				j = 0
				while j < n
					fxx = (x)
					hessian[i][j] = fxx
					j += 1
				end
				i += 1
			end
			return hessian

		end

		def Calculus.FindCriticalPoints(f, a, b, epsilon)
			criticalPoints = List[System::Double].new()
			# Evaluate function at endpoints of interval
			fa = Calculus.f(a)
			fb = Calculus.f(b)
			# Check if endpoints are critical points
			if Math.Abs(fa) < epsilon then
				criticalPoints.Add(a)
			end
			if Math.Abs(fb) < epsilon then
				criticalPoints.Add(b)
			end
			# Find critical points in the interval
			x = a
			fx = Calculus.f(a)
			while x < b
				xNext = x + epsilon
				fxNext = Calculus.f(xNext)
				# Check if the current point is a critical point
				if Math.Abs(fx) < epsilon and x > a and x < b then
					criticalPoints.Add(x)
				end
				# Check if the function changes sign between x and xNext
				if fx * fxNext < 0 then
					# Use bisection method to find a root
					xLeft = x
					xRight = xNext
					while xRight - xLeft > epsilon
						xMid = (xLeft + xRight) / 2
						fxMid = Calculus.f(xMid)
						if Math.Abs(fxMid) < epsilon then
							criticalPoints.Add(xMid)
							break
						end
						if fx * fxMid < 0 then
							xRight = xMid
						else
							xLeft = xMid
						end
					end
				end
				x = xNext
				fx = fxNext
			end
			return criticalPoints.ToArray()
		end

		class RiemannSumType
			def initialize()
			end
		end

		def RiemannSumType.RiemannSum(f, a, b, n, type)
			dx = (b - a) / n
			sum = 0
			i = 0
			while i < n
				x = a + i * dx
				fx = Calculus.f(x)
				case type
					when RiemannSumType.Left
						sum += fx * dx
					when RiemannSumType.Right
						sum += Calculus.f(x + dx) * dx
					when RiemannSumType.Midpoint
						sum += Calculus.f(x + dx / 2) * dx
					when RiemannSumType.Trapezoidal
						sum += (fx + Calculus.f(x + dx)) * dx / 2
				end
				i += 1
			end
			return sum
		end

		def RiemannSumType.DefiniteIntegral(f, a, b, n)
			return Calculus.RiemannSum(f, a, b, n, RiemannSumType.Midpoint)
		end
	end
end

File: calculus/calculus.cs
using System;
using System.Collections.Generic;

namespace Sigmath
{
    public class Calculus
    {
        public static double Derivative(Func<double, double> f, double x)
        {
            const double dx = 1e-9; // small increment to find the derivative
            double y1 = f(x);
            double y2 = f(x + dx);
            double derivative = (y2 - y1) / dx;
            return derivative;
        }

        public static double Integral(Func<double, double> f, double a, double b, int n)
        {
            double h = (b - a) / n;
            double integral = 0;
            for (int i = 1; i < n; i++)
            {
                double x = a + i * h;
                integral += f(x);
            }
            integral += (f(a) + f(b)) / 2;
            integral *= h;
            return integral;
        }

        public static double Limit(Func<double, double> f, Func<double, double> g, double x)
        {
            const int maxIterations = 1000;
            const double tolerance = 1e-9;
            double fx = f(x);
            double gx = g(x);
            double limit = fx / gx;
            for (int i = 0; i < maxIterations; i++)
            {
                double fx1 = Derivative(f, x);
                double gx1 = Derivative(g, x);
                double newLimit = fx1 / gx1;
                if (Basic.Abs(newLimit - limit) < tolerance)
                {
                    return newLimit;
                }
                limit = newLimit;
                x += 1e-9;
            }
            throw new Exception("Limit does not exist or cannot be computed with L'Hopital's Rule.");
        }

        public static double FindRoot(Func<double, double> f, double a, double b, double epsilon = 1e-6)
        {
            if (f(a) * f(b) > 0)
                throw new ArgumentException("Function must have opposite signs at interval endpoints");

            double c = (a + b) / 2;

            while (Math.Abs(f(c)) > epsilon)
            {
                if (f(a) * f(c) < 0)
                    b = c;
                else
                    a = c;

                c = (a + b) / 2;
            }

            return c;
        }

        private static int Factorial(int n)
        {
            int result = 1;

            for (int i = 2; i <= n; i++)
                result *= i;

            return result;
        }

        public static Func<double[], double> PartialDerivative(Func<double[], double> f, int index, double h = 1e-6)
        {
            return x =>
            {
                var xPlus = (double[])x.Clone();
                xPlus[index] += h;

                var xMinus = (double[])x.Clone();
                xMinus[index] -= h;

                return (f(xPlus) - f(xMinus)) / (2 * h);
            };
        }

        public static Func<double[], double[]> Gradient(Func<double[], double> f, double h = 1e-6)
        {
            return x =>
            {
                var gradient = new double[x.Length];

                for (int i = 0; i < x.Length; i++)
                    gradient[i] = PartialDerivative(f, i, h)(x);

                return gradient;
            };
        }

        public static Func<double[], double[,]> Hessian(Func<double[], double> f, double h = 1e-6)
        {
            return x =>
            {
                int n = x.Length;
                var hessian = new double[n, n];

                for (int i = 0; i < n; i++)
                {
                    for (int j = 0; j < n; j++)
                    {
                        var fxx = PartialDerivative(PartialDerivative(f, i, h), j, h)(x);
                        hessian[i, j] = fxx;
                    }
                }

                return hessian;
            };
        }

        public static double[] FindCriticalPoints(Func<double, double> f, double a, double b, double epsilon = 1e-6)
        {
            var criticalPoints = new List<double>();

            // Evaluate function at endpoints of interval
            double fa = f(a);
            double fb = f(b);

            // Check if endpoints are critical points
            if (Math.Abs(fa) < epsilon) criticalPoints.Add(a);
            if (Math.Abs(fb) < epsilon) criticalPoints.Add(b);

            // Find critical points in the interval
            double x = a;
            double fx = f(a);

            while (x < b)
            {
                double xNext = x + epsilon;
                double fxNext = f(xNext);

                // Check if the current point is a critical point
                if (Math.Abs(fx) < epsilon && x > a && x < b)
                    criticalPoints.Add(x);

                // Check if the function changes sign between x and xNext
                if (fx * fxNext < 0)
                {
                    // Use bisection method to find a root
                    double xLeft = x;
                    double xRight = xNext;

                    while (xRight - xLeft > epsilon)
                    {
                        double xMid = (xLeft + xRight) / 2;
                        double fxMid = f(xMid);

                        if (Math.Abs(fxMid) < epsilon)
                        {
                            criticalPoints.Add(xMid);
                            break;
                        }

                        if (fx * fxMid < 0)
                            xRight = xMid;
                        else
                            xLeft = xMid;
                    }
                }

                x = xNext;
                fx = fxNext;
            }

            return criticalPoints.ToArray();
        }

        public enum RiemannSumType
        {
            Left,
            Right,
            Midpoint,
            Trapezoidal
        }

        public static double RiemannSum(Func<double, double> f, double a, double b, int n, RiemannSumType type = RiemannSumType.Left)
        {
            double dx = (b - a) / n;
            double sum = 0;

            for (int i = 0; i < n; i++)
            {
                double x = a + i * dx;
                double fx = f(x);

                switch (type)
                {
                    case RiemannSumType.Left:
                        sum += fx * dx;
                        break;
                    case RiemannSumType.Right:
                        sum += f(x + dx) * dx;
                        break;
                    case RiemannSumType.Midpoint:
                        sum += f(x + dx / 2) * dx;
                        break;
                    case RiemannSumType.Trapezoidal:
                        sum += (fx + f(x + dx)) * dx / 2;
                        break;
                }
            }

            return sum;
        }

        public static double DefiniteIntegral(Func<double, double> f, double a, double b, int n = 1000)
        {
            return RiemannSum(f, a, b, n, RiemannSumType.Midpoint);
        }
    }
}
